util/vec-map/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Provides a map (key-value store) interface backed by `heapless::Vec`
+
+#![deny(missing_docs)]
+
+use displaydoc::Display;
+use heapless::Vec;
+
+/// An error which can occur when using VecMap
+#[derive(Clone, Debug, Display)]
+pub enum Error {
+    /// VecMap capacity exceeded
+    CapacityExceeded,
+}
+
+/// This is a mini version of VecMap that is no_std compatible and uses
+/// heapless::Vec instead alloc::Vec.
+///
+/// It may be better to patch upstream for no_std compatibility and use that,
+/// but that crate has other issues -- it relies on an
+/// experimental "contracts" crate that causes a dependency on rand crate.
+/// Porting to Heapless would be a breaking chnage to the API.
+///
+/// TBD what the best path is: https://github.com/p-avital/vec-map-rs/blob/master/src/lib.rs
+#[derive(Clone, Debug)]
+pub struct VecMap<K: Clone + Eq + PartialEq, V, const N: usize> {
+    keys: Vec<K, N>,
+    values: Vec<V, N>,
+}
+
+impl<K: Clone + Eq + PartialEq, V, const N: usize> Default for VecMap<K, V, N> {
+    fn default() -> Self {
+        Self {
+            keys: Default::default(),
+            values: Default::default(),
+        }
+    }
+}
+
+impl<K: Clone + Eq + PartialEq, V, const N: usize> VecMap<K, V, N> {
+    /// Check if the map is empty
+    #[inline]
```

Comment:
I think putting `#[inline]` on `fn is_empty()` should always be correct -- the good codegen is to load a usize and compare it with zero. Any time you are calling `is_empty()` you will probably also need that usize for a subsequent call. This is practically the poster child for putting `#[inline]` on a function.

If anything I would expect this to reduce compile times, since we're allowing the compiler to figure out earlier in the process that this should be inlined, as opposed to waiting for some LTO step where more context has been lost and a bunch of other code has already been generated.

Reply:
We've posted references for when to use `#[inline]` elsewhere in slack i think, probably in the #rust-dev channel

> Premature inlining has some downsides:

I think your first 4 points simply don't apply in this case.

The fifth one, I take issue with:

>     over inlining can cause it's own run time performance degradation.

In general, you will get the best run time performance by letting the compiler decide whether to inline something or not.

It is important to understand what the `#[inline]` directive does -- it ALLOWS the compiler to inline a function across crate boundaries, if it thinks that there will be an improvement. It does not force the compiler to inline things. In general the compiler is a better judge than you are of whether this is good on a particular architecture.

There is a different directive `#[inline(always)]` that is different and should be used sparingly.

Reply:
I think it sounds reasonable, I saw that the underlying `is_empty()` and `len()` for the backend types were also inlined so you end up loosing that here, unless you opt into LTO. I've just got some baggage from inline mania causing more harm than good :smile:

Reply:
> We've posted references for when to use #[inline] elsewhere in slack i think

I'm only finding your and my discussion about compile times in `#rust-devs` for references to inline

> In general the compiler is a better judge than you are of whether this is good on a particular architecture.

I think the often overlooked thing is that the `inline` keyword/attribute nudges the compiler more than it would have been, even with LTO on.  Many optimizations are hueristics and it will take into account that you said you thought it would be good into weighting that decision.

> The only reason not to mark things #[inline] is:

I don't think you're saying inline everything unless, but it's very easy for people to think that from the remark.  I'll have to keep looking, even in c/c++ the `inline` keyword is a hint to the compiler to inline across libraries.